### ISSUE TITLE
Setting YII_DEBUG=false in humhub 1.15.2 leads to internal server error

### ISSUE THREAD

**Issue by ukaeufe** [humhub/humhub#6846]
### What steps will reproduce the problem?

Setting YII_DEBUG=false in index.php in humhub root directory quickly leads to internal server error.

### What is the expected result?

Humhub should operate normal and in production mode.

### What do you get instead?

Internal server error, which is resolved immediately when setting YII_DEBUG back to true.

### Additional info

| Q                | A
| ---------------- | ---
| humhub     | 1.15.2
| Yii version      | Yii v2.0.50-dev
| PHP version      |  8.2
| Operating system | Linux ae93c 5.10.0-26-amd64 #1 SMP Debian 5.10.197-1 (2023-09-29) x86_64 GNU/Linux
| Env | Netcup webhosting / plesk

**Comment by luke-** [comment]
What kind of error message are you getting exactly?

Have you commented out the lines like here?
https://docs.humhub.org/docs/admin/installation#enable-production-mode

**Comment by ukaeufe** [comment]
I get a lot of error messages that most likely are just side effects. I followed the instructions to enable prod mode and ran into this problem. I narrowed down the issue that the internal server error appears when setting YII_DEBUG=false.
I wonder how can this have such an effect? YII_DEBUG shall only have impact to the logging level and not change the app behaviour. This is either a humhub or a yii2 issue. Also I see that humhub 1.15.2 uses a dev drop of yii2 which in my opinion should be replaced by the last official version . 

 I get errors related to openbase_dir . There a specific module is mentioned e.g. jitsi_meet. In another installation on the same Webhoster I get a similar error for another module. 

`yii\base\ErrorException: is_file(): open_basedir restriction in effect. File(/connect.duesseldorf-vegan.de/httpdocs/protected/modules/jitsi-meet/Events.php) is not within the allowed path(s): (/var/www/vhosts/hosting196037.ae93c.netcup.net/:/tmp/:/var/lib/php/sessions) in /var/www/vhosts/hosting196037.ae93c.netcup.net/connect.duesseldorf-vegan.de/httpdocs/protected/vendor/yiisoft/yii2/BaseYii.php:286
Stack trace:
#0 [internal function]: yii\base\ErrorHandler->handleError()
#1 /var/www/vhosts/hosting196037.ae93c.netcup.net/connect.duesseldorf-vegan.de/httpdocs/protected/vendor/yiisoft/yii2/BaseYii.php(286): is_file()
#2 [internal function]: yii\BaseYii::autoload()
#3 /var/www/vhosts/hosting196037.ae93c.netcup.net/connect.duesseldorf-vegan.de/httpdocs/protected/humhub/components/ModuleManager.php(223): method_exists()
#4 /var/www/vhosts/hosting196037.ae93c.netcup.net/connect.duesseldorf-vegan.de/httpdocs/protected/humhub/components/ModuleManager.php(136): humhub\components\ModuleManager->register()
#5 /var/www/vhosts/hosting196037.ae93c.netcup.net/connect.duesseldorf-vegan.de/httpdocs/protected/humhub/components/bootstrap/ModuleAutoLoader.php(35): humhub\components\ModuleManager->registerBulk()
#6 /var/www/vhosts/hosting196037.ae93c.netcup.net/connect.duesseldorf-vegan.de/httpdocs/protected/vendor/yiisoft/yii2/base/Application.php(325): humhub\components\bootstrap\ModuleAutoLoader->bootstrap()
#7 /var/www/vhosts/hosting196037.ae93c.netcup.net/connect.duesseldorf-vegan.de/httpdocs/protected/vendor/yiisoft/yii2/web/Application.php(69): yii\base\Application->bootstrap()
#8 /var/www/vhosts/hosting196037.ae93c.netcup.net/connect.duesseldorf-vegan.de/httpdocs/protected/humhub/components/Application.php(61): yii\web\Application->bootstrap()
#9 /var/www/vhosts/hosting196037.ae93c.netcup.net/connect.duesseldorf-vegan.de/httpdocs/protected/vendor/yiisoft/yii2/base/Application.php(271): humhub\components\Application->bootstrap()
#10 /var/www/vhosts/hosting196037.ae93c.netcup.net/connect.duesseldorf-vegan.de/httpdocs/protected/humhub/components/Application.php(42): yii\base\Application->init()
#11 /var/www/vhosts/hosting196037.ae93c.netcup.net/connect.duesseldorf-vegan.de/httpdocs/protected/vendor/yiisoft/yii2/base/BaseObject.php(109): humhub\components\Application->init()
#12 /var/www/vhosts/hosting196037.ae93c.netcup.net/connect.duesseldorf-vegan.de/httpdocs/protected/vendor/yiisoft/yii2/base/Application.php(204): yii\base\BaseObject->__construct()
#13 /var/www/vhosts/hosting196037.ae93c.netcup.net/connect.duesseldorf-vegan.de/httpdocs/protected/humhub/components/ApplicationTrait.php(38): yii\base\Application->__construct()
#14 /var/www/vhosts/hosting196037.ae93c.netcup.net/connect.duesseldorf-vegan.de/httpdocs/index.php(27): humhub\components\Application->__construct()
#15 {main}`

**Comment by ukaeufe** [comment]
just updated to 1.15.3 same issue

**Comment by ArchBlood** [comment]
Overall this is a similar issues as #5155;

It seems like your HumHub is encountering an "`open_basedir` restriction" error, indicating that the file it's trying to access is not within the allowed paths specified by the `open_basedir` directive in your PHP configuration.

The specific file causing the issue is `/connect.duesseldorf-vegan.de/httpdocs/protected/modules/jitsi-meet/Events.php`. The error occurs because this file is located outside of the allowed paths (`/var/www/vhosts/hosting196037.ae93c.netcup.net/:/tmp/:/var/lib/php/sessions`).

To resolve this issue, you have a few options:

Adjust `open_basedir` settings: You can modify the `open_basedir` directive in your PHP configuration to include the directory where the file is located. However, be cautious when making changes to this directive as it affects the security of your server.

Move the file: If possible, move the `Events.php` file to a location within the allowed paths specified by `open_basedir`.

Consult with your hosting provider: If you're not familiar with server configuration, it might be best to consult with your hosting provider or a server administrator for assistance in resolving this issue safely.

Once you've addressed the `open_basedir` restriction, the error should no longer occur.

**Comment by ukaeufe** [comment]
Thx a lot. I may go acc to your proposal. At the same time, I think this is fixing the symptoms and not the root cause.
Think about it. I just change a setting that should impact the level of logging. How can such a setting impact the overall behaviour of the php app including suddenly not being able to access php script that are accessible when YII_DEBUG is set to true. The jitsi_meet module if fully functional. I can do virtual meetings with it. Also in a parallel installation of humhub I have the same issue and another module is mentioned in the open_basedir error.
So in summary, I assume the root cause is in the way YII_DEBUG impacts the humhub application. This should be checked. As most of the magic happens in the yii2 framework, I opened the same issue there too.

**Comment by rob006** [comment]
> Think about it. I just change a setting that should impact the level of logging.

`YII_DEBUG` does a lot more than logging level.

Note difference between absolute paths: Yii tries to load file in `/connect.duesseldorf-vegan.de/httpdocs/`, but your app directory is `/var/www/vhosts/hosting196037.ae93c.netcup.net/connect.duesseldorf-vegan.de/httpdocs/`. If both paths point to the same directory, then open_basedir restrictions should be updated and contain both paths. In other case there is something wrong with aliases.

**Comment by ukaeufe** [comment]
Yes makes sense . Still I would like to understand how the path issue relates to you debug . Will try to check where these paths are defined

**Comment by ArchBlood** [comment]
> Yes makes sense . Still I would like to understand how the path issue relates to you debug . Will try to check where these paths are defined

The error you're encountering, specifically the `open_basedir` restriction issue, is primarily a server configuration issue rather than an application-specific one.

`open_basedir` is a PHP configuration directive that restricts the files PHP can access based on their path. When PHP tries to access a file outside of the allowed paths specified by `open_basedir`, it raises an error, as you've experienced.

In your case, the error indicates that the file `Events.php` within the `jitsi-meet` module is not within the allowed paths specified by `open_basedir`. This suggests that the server's PHP configuration is too restrictive and does not include the necessary directories for your HumHub to function properly as @rob006 stated.

To resolve this issue, you'll need to adjust the `open_basedir` directive in your server's PHP configuration to include the necessary directories for your HumHub, such as the directory containing the `Events.php` file for the `jitsi-meet` module.

Once you update the server configuration to include the required paths, PHP will be able to access the necessary files, and the error should no longer occur.

**Comment by rob006** [comment]
> Yes makes sense . Still I would like to understand how the path issue relates to you debug . Will try to check where these paths are defined

I'm not familiar with HumHub, but debug mode sometimes enables additional modules (which may affect code execution order and thus autoloading) or alter behavior of some modules (disabling cache is quite common). Even HumHub has something like this:

https://github.com/humhub/humhub/blob/f8c5cfaf8417aba84d5c50940dd8e061009f3886/protected/humhub/components/bootstrap/ModuleAutoLoader.php#L47-L50

**Comment by ArchBlood** [comment]
I believe a lot of this should be fixed within #6816, and my https://github.com/humhub/humhub/pull/6816#discussion_r1495067906 should also further help with other possible issues that could happen.

Another approach could be the following based on #6816;

```php
<?php

namespace humhub\components\bootstrap;

use humhub\components\Application;
use Yii;
use yii\base\BootstrapInterface;
use yii\base\ErrorException;
use yii\base\InvalidArgumentException;
use yii\base\InvalidConfigException;
use yii\helpers\FileHelper;
use yii\caching\CacheInterface;

class ModuleAutoLoader implements BootstrapInterface
{
    public const CACHE_ID = 'module_configs';
    public const CONFIGURATION_FILE = 'config.php';
    public const CACHE_EXPIRATION_TIME = 3600; // Cache expiration time in seconds (1 hour)

    public function bootstrap($app)
    {
        $modules = self::locateModules(Yii::$app->cache, Yii::$app->params['moduleAutoloadPaths']);
        Yii::$app->moduleManager->registerBulk($modules);
    }

    public static function locateModules(CacheInterface $cache, array $moduleAutoloadPaths): array
    {
        $modules = $cache->get(self::CACHE_ID);

        if ($modules === false || self::shouldRefreshCache($cache)) {
            $modules = static::findModules($moduleAutoloadPaths);
            $cache->set(self::CACHE_ID, $modules);
            $cache->set(self::CACHE_ID . '_timestamp', time());
        }

        return $modules;
    }

    private static function shouldRefreshCache(CacheInterface $cache): bool
    {
        if (YII_DEBUG) {
            return true;
        }
        
        $cacheTime = $cache->get(self::CACHE_ID . '_timestamp');
        if ($cacheTime === false || time() - $cacheTime > self::CACHE_EXPIRATION_TIME || self::cacheExpired($cache)) {
            return true;
        }
        
        return false;
    }

    private static function cacheExpired(CacheInterface $cache): bool
    {
        $cachedModules = $cache->get(self::CACHE_ID);
        
        if ($cachedModules === false) {
            return true;
        }
        
        foreach ($cachedModules as $module) {
            $modulePath = Yii::getAlias('@app') . DIRECTORY_SEPARATOR . $module['path'];
            if (file_exists($modulePath)) {
                $lastModified = filemtime($modulePath);
                if ($lastModified > $cache->get(self::CACHE_ID . '_timestamp')) {
                    return true;
                }
            } else {
                return true;
            }
        }
        
        return false;
    }

    private static function findModules(array $paths): array
    {
        $folders = [];
        foreach ($paths as $path) {
            try {
                $folders = array_merge($folders, self::findModulesByPath($path));
            } catch (InvalidArgumentException $ex) {
                throw new ErrorException('Invalid module autoload path: ' . $path);
            }
        }

        $modules = [];
        $moduleIdFolders = [];
        $preventDuplicatedModules = Yii::$app->moduleManager->preventDuplicatedModules;

        foreach ($folders as $folder) {
            try {
                $moduleConfig = static::getModuleConfigByPath($folder);
                if ($preventDuplicatedModules && isset($moduleIdFolders[$moduleConfig['id']])) {
                    Yii::error('Duplicated module "' . $moduleConfig['id'] . '"(' . $folder . ') is already loaded from the folder "' . $moduleIdFolders[$moduleConfig['id']] . '"');
                } else {
                    $modules[$folder] = $moduleConfig;
                    $moduleIdFolders[$moduleConfig['id']] = $folder;
                }
            } catch (\Throwable $e) {
                Yii::error($e);
            }
        }

        if ($preventDuplicatedModules) {
            foreach (Yii::$app->moduleManager->overwriteModuleBasePath as $overwriteModuleId => $overwriteModulePath) {
                if (isset($moduleIdFolders[$overwriteModuleId]) && $moduleIdFolders[$overwriteModuleId] !== $overwriteModulePath) {
                    try {
                        $moduleConfig = static::getModuleConfigByPath($overwriteModulePath);

                        Yii::info('Overwrite path of the module "' . $overwriteModuleId . '" to the folder "' . $overwriteModulePath . '"');
                        unset($modules[$moduleIdFolders[$overwriteModuleId]]);
                        $modules[$overwriteModulePath] = $moduleConfig;
                        $moduleIdFolders[$overwriteModuleId] = $overwriteModulePath;
                    } catch (\Throwable $e) {
                        Yii::error($e);
                    }
                }
            }
        }

        return $modules;
    }

    private static function getModuleConfigByPath(string $modulePath): array
    {
        $configFilePath = $modulePath . DIRECTORY_SEPARATOR . self::CONFIGURATION_FILE;
        if (!file_exists($configFilePath)) {
            throw new \RuntimeException('Configuration file not found: ' . $configFilePath);
        }

        $config = include $configFilePath;
        if (!is_array($config)) {
            throw new \RuntimeException('Invalid configuration file format: ' . $configFilePath);
        }

        return $config;
    }

    private static function findModulesByPath(string $path): array
    {
        $hasConfigurationFile = static function ($path) {
            return is_file($path . DIRECTORY_SEPARATOR . self::CONFIGURATION_FILE);
        };

        return FileHelper::findDirectories(
            Yii::getAlias($path, true),
            ['filter' => $hasConfigurationFile, 'recursive' => false]
        );
    }
}
```

**Comment by ukaeufe** [comment]
Thx @ArchBlood I will try this on my parallel test installation.

**Comment by ukaeufe** [comment]
Hi @ArchBlood I tried above patch by replacing the file content and then trying to activate production mode. Result is the same. Unfortunately, I lack the time to debug this further. So I will live with debug mode being active for now.

**Comment by ArchBlood** [comment]
> Hi @ArchBlood I tried above patch by replacing the file content and then trying to activate production mode. Result is the same. Unfortunately, I lack the time to debug this further. So I will live with debug mode being active for now.

Hmm, it could be because `YII_DEBUG` isn't being defined, and in this case you'd have to use `defined('YII_DEBUG') && YII_DEBUG` so something like the following would have to be changed in my modification;
```php
private static function shouldRefreshCache(CacheInterface $cache): bool
{
    if (defined('YII_DEBUG') && YII_DEBUG) {
        return true;
    }
    
    $cacheTime = $cache->get(self::CACHE_ID . '_timestamp');
    if ($cacheTime === false || time() - $cacheTime > self::CACHE_EXPIRATION_TIME || self::cacheExpired($cache)) {
        return true;
    }
    
    return false;
}
```
@luke- what would your opinion be on this?
